internal/bcdb/worldstate_query_processor.go

```diff
@@ -140,3 +143,58 @@ func (q *worldstateQueryProcessor) getNodeConfig(nodeID string) (*types.GetNodeC
 
 	return c, nil
 }
+
+func (q *worldstateQueryProcessor) executeJSONQuery(dbName, querierUserID string, query []byte) (*types.DataQueryResponse, error) {
+	if worldstate.IsSystemDB(dbName) {
+		return nil, &errors.PermissionErr{
+			ErrMsg: "no user can directly read from a system database [" + dbName + "]. " +
+				"To read from a system database, use /config, /user, /db rest endpoints instead of /data",
+		}
+	}
+
+	hasPerm, err := q.identityQuerier.HasReadAccessOnDataDB(querierUserID, dbName)
+	if err != nil {
+		return nil, err
+	}
+	if !hasPerm {
+		return nil, &errors.PermissionErr{
+			ErrMsg: "the user [" + querierUserID + "] has no permission to read from database [" + dbName + "]",
+		}
+	}
+
+	keys, err := q.jsonQueryExecutor.ExecuteQuery(dbName, query)
+	if err != nil {
+		return nil, err
+	}
+
+	var results []*types.KVWithMetadata
+
+	for k := range keys {
+		value, metadata, err := q.db.Get(dbName, k)
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO: we can store the ACL as value in the indexEntry. With that, we can avoid reading the whole value
+		// to perform the access control - issue #152
+		acl := metadata.GetAccessControl()
+		if acl != nil {
+			if !acl.ReadUsers[querierUserID] && !acl.ReadWriteUsers[querierUserID] {
+				continue
+			}
+		}
+
+		results = append(
```

Comment:
TODO 
- about controlling the returned number of keys, paging, etc... 
- controlling the size in bytes
- skipping, etc

Reply:
skipping is a feature. Not a todo and it requires a story. For others, we have a story in terms of restricting the response size (https://github.com/IBM-Blockchain/bcdb-server/issues/182) and memory management (https://github.com/IBM-Blockchain/bcdb-server/issues/174).  As I am working on those currently, I do not see value in putting a TODO now.

Reply:
I agree, by TODO I meant have a story, not necessarily a code comment. Just using the opportunity to flash out issues so we won't forget.